torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -347,6 +355,260 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+        (long long)shape[shape_idx], (long long)shape_idx, (long long)seq_len);
+    }
+    PyObject** items = PySequence_Fast_ITEMS(seq.get());
+    for (Py_ssize_t i = 0; i < seq_len; ++i) {
+      if (!_extract_ndarrays(items[i], shape, numels, shape_idx+1, array_ptr_storage)) {
+        return false;
+      }
+    }
+  } else {
+    return false;
+  }
+  return true;
+}
+
+
+bool extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  A function for extracting all numpy arrays embedded (recursively) within PySequences in obj. Failure to do so will return
+  false, else true. The extracted numpy array pointers are stored in array_ptr_storage. It uses _extract_ndarrays() function 
+  for the extraction. See the comments there. This function is used in internal_new_from_data() in "torch/csr/utils/tensor_new.cpp"
+  */
+
+  int64_t ndims = shape.size();
+  if (ndims == 0) {
+    return false;
+  } else if (shape[0] == 0) {
+    return false;
+  }
+
+  std::vector<int64_t> numels(ndims);
+  numels[ndims-1] = shape[ndims-1];
+
+  for (int64_t i = ndims-2; i >= 0; --i) {
+    numels[i] = shape[i] * numels[i+1];
+  }
+
+  return _extract_ndarrays(obj, shape, numels, 0, array_ptr_storage);
+}
+
+
+template <typename T>
+inline void memcpy_cpu(void* dst, const void* src, size_t num_items, bool isaligned, int itemsize=1) {
+  /*
+  A utility function for copying an array buffer of type T from src to dst using parallel loops.
+  This is used in the function tensor_from_ndarray_batch below for copying raw memory from 
+  (contiguous) numpy arrays to an output pytorch tensor.
+  */
+  if (isaligned) {
+    at::parallel_for(
+      0,
+      num_items,
+      at::internal::GRAIN_SIZE,
+      [=](size_t begin, size_t end) {  
+        T* src_tmp = ((T*) src) + begin;
+        T* dst_tmp = ((T*) dst) + begin;
+        auto last = src_tmp+(end-begin);
+
+        while(src_tmp < last){
+          *dst_tmp = *src_tmp;
+          ++src_tmp;
+          ++dst_tmp;
+        }
+      });
+  } 
+  else {
+    at::parallel_for(
+      0,
+      num_items*itemsize,
+      at::internal::GRAIN_SIZE,
+      [=](size_t begin, size_t end) {  
+        char* src_tmp = ((char*) src) + begin;
+        char* dst_tmp = ((char*) dst) + begin;
+        auto last = src_tmp+(end-begin);
+
+        while(src_tmp < last){
+          *dst_tmp = *src_tmp;
+          ++src_tmp;
+          ++dst_tmp;
+        }
+      });    
+  }
+}
+
+
+inline void* store_ndarray(void* tensor_ptr, void* ndarray_ptr, at::ScalarType scalarType, size_t num_items, int itemsize, bool isaligned) {
+  /* 
+  This utility function is used for copying elements from a raw numpy array memory to a tensor memory location by using memcpy_cpu and proper type matching.
+  */
+  switch (scalarType) {
+    case at::kByte: memcpy_cpu<uint8_t> (tensor_ptr, ndarray_ptr, num_items, itemsize, isaligned); return (void*)(((uint8_t *)tensor_ptr) + num_items);
```

Comment:
you should be able to use dispatch macro here instead of spelling out all the dtypes. Bonus points for using OpaqueType (see in ATen/native/cuda/IndexKernel.cu) to minimize number of template instantiations, but the benefit may be too small to bother.

Reply:
Could you please refer me to where I can find this macro, or any source file where it has been used? Still getting used to the codebase.

Reply:
For an example use see e.g. here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/IndexKernel.cu#L210. Dispatch macro defines C type `scalar_t` depending on scalarType, that you can use to instantiate templated functions, it is implemented pretty much like this switch statement that you have here.